waybar: init at 0.4.0

FlorianFranzen opened the pull request:
###### Motivation for this change

Initial addition of [waybar](https://github.com/Alexays/Waybar).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

FlorianFranzen committed:
waybar: init at 0.4.0

alyssais commented on pkgs/applications/misc/waybar/default.nix:
```diff
@@ -0,0 +1,54 @@
+{ 
```
We don’t usually leave a newline here

alyssais commented on pkgs/applications/misc/waybar/default.nix:
```diff
@@ -0,0 +1,54 @@
+assert swaySupport -> sway != null;
+
+stdenv.mkDerivation rec {
+    name = "waybar-${version}";
```
Nix code should be indented with two spaces.

alyssais commented on pkgs/applications/misc/waybar/default.nix:
```diff
@@ -0,0 +1,54 @@
+
+    nativeBuildInputs = [
+      meson ninja pkgconfig
+    ];
```
Perhaps put this next to the `buildInputs`. No need for it to be more than one line, either.

alyssais commented on pkgs/applications/misc/waybar/default.nix:
```diff
@@ -0,0 +1,54 @@
+      ++ stdenv.lib.optional  traySupport  libdbusmenu-gtk3
+      ++ stdenv.lib.optional  pulseSupport libpulseaudio
+      ++ stdenv.lib.optional  nlSupport    libnl
+      ++ stdenv.lib.optionals swaySupport  [ sway makeWrapper ];
```
You can take `lib` as an input and save yourself writing `stdenv.` over and over.

alyssais commented on pkgs/applications/misc/waybar/default.nix:
```diff
@@ -0,0 +1,54 @@
+assert traySupport -> libdbusmenu-gtk3 != null;
+assert pulseSupport  -> libpulseaudio != null;
+assert nlSupport -> libnl != null;
+assert swaySupport -> sway != null;
```
There’s no need for this. You don’t need to provide default values for the package parameters.

alyssais commented on pkgs/applications/misc/waybar/default.nix:
```diff
@@ -0,0 +1,54 @@
+    postInstall = if swaySupport then ''
+      wrapProgram $out/bin/waybar \
+        --prefix PATH : "${sway}/bin"
+    '' else "";
```
You can use `lib.optionalString` here.

alyssais commented on pkgs/applications/misc/waybar/default.nix:
```diff
@@ -0,0 +1,54 @@
+
+    postInstall = if swaySupport then ''
+      wrapProgram $out/bin/waybar \
+        --prefix PATH : "${sway}/bin"
```
Can you explain what this does? Does waybar use the `sway` executable?

FlorianFranzen replied:
This is no longer needed, at least with sway-beta. The latest version of the bar does not work with sway < 1.0 anyway, so I dropped the the ```postInstall``` step completely.

alyssais reviewed (changes requested):
Lots of comments, but don’t worry. Mostly formatting.

jtojnar commented on pkgs/applications/misc/waybar/sysconfdir.patch:
```diff
@@ -0,0 +1,13 @@
+ install_data(
+     './resources/config',
+     './resources/style.css',
+-    install_dir: join_paths(get_option('out'), 'etc/xdg/waybar')
```
This is actually correct and we would want [more applications](https://github.com/hughsie/fwupd/pull/824) to do this.

You should add `"-Dout=${placeholder "out"}"` and `"--sysconfdir=/etc"` to `mesonFlags`.

FlorianFranzen replied:
Thanks for the hint. Why do I need ```"--sysconfdir=/etc"```? Unless the ```sysconfdir``` is passed by meson to the applications, settings this should not make a difference.

jtojnar replied:
Yup, in this case sysconfdir will probably have no effect, as they use `g_get_system_data_dirs` to find the configuration.

FlorianFranzen commented:
Update package file as requested, see rewritten commit.

alyssais commented on pkgs/applications/misc/waybar/default.nix:
```diff
@@ -0,0 +1,42 @@
+{ stdenv, fetchFromGitHub, meson, pkgconfig, ninja, 
+  gtkmm3, jsoncpp, libinput, libsigcxx, wayland, wlroots, fmt, 
```
Commas go at the start of the following line, rather than the end of the line.

alyssais commented on pkgs/applications/misc/waybar/default.nix:
```diff
@@ -0,0 +1,42 @@
+}:
+  stdenv.mkDerivation rec {
+    name = "waybar-${version}";
+    version = "0.3.0";
```
```suggestion
    version = "0.3.0";

```

alyssais commented on pkgs/applications/misc/waybar/default.nix:
```diff
@@ -0,0 +1,42 @@
+    ];
+
+    meta = with stdenv.lib; {
+      description = "Highly customizable Wayland bar for Sway and Wlroots based compositors.";
```
```suggestion
      description = "Highly customizable Wayland bar for Sway and Wlroots based compositors";
```

From the [manual](https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes):

>  Don’t include a period at the end. Don’t include newline characters. Capitalise the first character. For brevity, don’t repeat the name of package — just describe what it does.

alyssais reviewed (changes requested):
Couple more style notes.

alyssais commented on pkgs/applications/misc/waybar/default.nix:
```diff
@@ -0,0 +1,43 @@
+, traySupport  ? true,  libdbusmenu-gtk3
+, pulseSupport ? false, libpulseaudio
+, nlSupport    ? true,  libnl
+, swaySupport  ? false, sway-beta
```
What's the closure size impact of just defaulting all of these to true?

FlorianFranzen replied:
According to ```nix path-info -S ...```, this would change the size from 287,863,928 to 294,567,016.

alyssais replied:
That's not a big difference. Do you think we should just enable all of them?

alyssais commented on pkgs/top-level/all-packages.nix:
```diff
@@ -17571,11 +17571,15 @@ in
     stPatches = config.st.patches or null;
   };
-
+ 
```
```suggestion

```

FlorianFranzen commented:
@alyssais Now trailing whitespace free.

FlorianFranzen committed:
nixos/waybar: init module

FlorianFranzen commented:
Now also includes a NixOS module to run waybar as systemd service. 

@primeos I am interested in implement sway's recommed [systemd integration](https://github.com/swaywm/sway/wiki/Systemd-integration) in NixOS. Any feedback to this module would be welcome, especially on how to pass the ```SWAYSOCK``` variable to waybay and others (without a call to ```systemctl --user import-environment ...```).

FlorianFranzen commented:
@alyssais Is there anything else missing before we can get this merged?

primeos commented:
> @primeos I am interested in implement sway's recommed [systemd integration](https://github.com/swaywm/sway/wiki/Systemd-integration) in NixOS. Any feedback to this module would be welcome, especially on how to pass the `SWAYSOCK` variable to waybay and others (without a call to `systemctl --user import-environment ...`).

Sorry for my late response, I'm not a huge fan of systemd user units (at least ATM) and therefore also not that familiar with them.

There should be multiple ways to pass `SWAYSOCK` but I'm not sure which one is the best. It would be nice if `SWAYSOCK` could be inherited by having a dependency on `sway.service` but I assume this isn't the case. It should be possible to use `systemctl --user import-environment SWAYSOCK` from withing `sway.service` but that solution could get a bit complicated (doesn't seem like the worst idea though). Another approach would be to hard-code `SWAYSOCK` but this will become problematic if users need multiple Sway sessions (though that most likely won't work well with systemd anyway) and might cause other problems. Using systemd environment file generators (`man systemd.environment-generator`) should also be possible for a more dynamic approach.

But in any case it might be best to first check how other units are solving this problem.

alyssais commented:
@FlorianFranzen I'm happy to merge this once I see an ack of @primeos's comment from you. Let me know if you plan to change it further – I was about to merge it before and then a whole new module appeared!

FlorianFranzen commented:
In do not want to make ```waybar``` sway-only as it is not by design. That is why I made it depend on the ```graphical-session.target``` instead, which should be the correct and general way (see e.g. #22671).

@primeos Thank you for your feedback. As ```SWAYSOCK``` is needed inside a sway session for the IPC message client to function anyway, any systemd solution will have to take care of this as well. If I find some time to clean up my solution, I will make sure to open a pull request and let you know.

Hardcoding ```SWAYSOCK``` is not sensible solution as the name always contains the process id of sway (i.e. ```/run/user/1000/sway-ipc.1000.1547.sock```).

@alyssais I have no plans to change this further, thank you for your help to get this merged.

FlorianFranzen commented:
Updated waybar to latest release.

alyssais commented:
Sorry for how slow I've been on this.

I just took a look, and it needs to be updated, because with sway 1.0,
there's no longer a sway-beta package.

FlorianFranzen commented:
@alyssais: No problem, that was bound to happen. I rebased the pull request against the latest unstable and replaced all reference to ```sway-beta``` accordingly.

alyssais commented:
Thank you for your persistence!
